[WIP] Try: Add sticky position support for the Group block

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

The goal is to solve #30121 

🚧 🚧 🚧 🚧 🚧 🚧 

This is a WIP exploring adding Position support to allow a group block to be sticky positioned, details TBC.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

TBC

## Screenshots <!-- if applicable -->

Work in progress:

![Kapture 2022-01-18 at 17 42 09](https://user-images.githubusercontent.com/14988353/149884772-a2dad9b2-9365-4739-a1bb-044c9353ba17.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ramonjd commented on lib/compat/wordpress-5.9/get-global-styles-and-settings.php:
```diff
@@ -23,7 +23,7 @@
 	 * @return array The settings to retrieve.
 	 */
-	function wp_get_global_settings( $path = array(), $context = array() ) {
+	function gutenberg_get_global_settings( $path = array(), $context = array() ) {
```
Thanks for getting this up so we can have a looksy @andrewserong !!

Looks like a similar approach to the one we suggested over at https://github.com/WordPress/gutenberg/pull/38671#issuecomment-1034433841 and over at https://github.com/WordPress/gutenberg/pull/38816/files#r808663126, where I'm looking to tackle something similar.

As mentioned over at https://github.com/WordPress/gutenberg/pull/38816/files#r808663126, I'm starting to wonder if we should move changes to this file over to `compat/6.0` in a "compatibility preparation branch or something" since we'll have perform some further trickery on it.

By "trickery" I mean updating methods such as `{wp|gutenberg}_get_global_stylesheet()` to use `WP_Theme_JSON_Resolver_Gutenberg` instead of `WP_Theme_JSON_Resolver`, so it picks up the compat/6.0 changes in `WP_Theme_JSON_Gutenberg` for example.

I'm still lacking a lot of the context and knowledge of the approaches that were taken for 5.9, so maybe @oandregal @jorgefilipecosta @aristath or @youknowriad might have some tips for us.

oandregal replied:
I'm looking at this one and I'm going to push changes in the next few hours.

oandregal replied:
Hey Andrew, I've heavily edited the commits in this PR to show how can we override global styles code. Sorry about the intrusion. :bowing_man: 

Here's what we can do to add new things to the global styles code in the server:

1. Update the allowed lists (`PROPERTIES_METADATA`, `VALID_SETTINGS`, etc) with the new settings in `lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php`. This suffices to make the plugin work in WordPress 5.8. See this commit for example https://github.com/WordPress/gutenberg/pull/38039/commits/e2a18908c1071490917fb437a73bcd01f9f0b814
2. Make sure the plugin calls the plugin classes (`WP_Theme_JSON_Resolver_Gutenberg` and `WP_Theme_JSON_Gutenberg`). Because our main interfaces are the `wp_get_global_*` functions, the changes in step 1 are not enough for WordPress 5.9 because those functions are part of core and call the core classes. In WordPress 5.8 it works fine because it uses the plugin classes. How can we override the functions in 5.9? We can't, so we instead call them `gutenberg_get_global_*` in the plugin and use those in all the instances. This commit renames `wp_get_global_settings` to `gutenberg_get_global_settings` and changes all the instances https://github.com/WordPress/gutenberg/pull/38039/commits/cf100a22e52d52bedcc9ecd021d872e6eb84d51b

That's it. Note that we only need the second step once: when this is done, subsequent PRs won't have to do it, they'll only require the changes in step 1.

Andrew, I've noticed you have added `layout.position` to the properties metadata, so I understood that'd be a valid style going forward. I went ahead I repeated the process I did with settings:

1. Update the allowed lists in `lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php`. https://github.com/WordPress/gutenberg/pull/38039/commits/3e935bd4c9571d2272c66b248de6b2d118f7a7fd
2. Make sure the plugin functions call the plugin classes: https://github.com/WordPress/gutenberg/pull/38039/commits/f20fa0c5262d1e8797da3dc377e7e349fa3e15d8 In testing this I've noticed a bug that I've fixed at https://github.com/WordPress/gutenberg/pull/38039/commits/ca41acfa4bea1655a590159563581c887a5467d6

oandregal replied:
I've figured there are a few in-flight PRs you all are working on and we don't know which will land first. So I've gone ahead and extracted the changes done in step 2 (both settings & styles) to a standalone PR so we can fast-track them and your PRs focus on feature work https://github.com/WordPress/gutenberg/pull/38883

Once that lands, this PR can be rebased and the related commits removed.

oandregal replied:
cc @andrewserong @ramonjd @aaronrobertshaw @ajlende @scruffian for awareness of this thread, as I know you all have been working on this area.

ramonjd replied:
Huge kudos for helping on this @oandregal! 🙇 And for the great explanations. It's a massive help. Thank you!!

andrewserong replied:
That's fantastic, thanks so much for jumping in and committing to the PR @oandregal, much appreciated! We'll take a look at #38883 today 😀

ramonjd commented on lib/compat/wordpress-5.9/class-wp-theme-json-5-9.php:
```diff
@@ -179,6 +179,7 @@ class WP_Theme_JSON_5_9 {
 		'padding-right'              => array( 'spacing', 'padding', 'right' ),
 		'padding-bottom'             => array( 'spacing', 'padding', 'bottom' ),
 		'padding-left'               => array( 'spacing', 'padding', 'left' ),
+		'position'                   => array( 'layout', 'position' ),
```
Should this be in the `compat/6.0/class-wp-theme-json-gutenberg.php`

Haven't checked if this branch has been rebased yet.

```php
class WP_Theme_JSON_Gutenberg extends WP_Theme_JSON_Gutenberg_5_9 {
	/**
	 * The valid properties under the settings key.
	 *
	 * @var array
	 */
	const VALID_SETTINGS = array(
		...
		'layout'          => array(
			'contentSize' => null,
			'position'    => null,
			'wideSize'    => null,
		),
		...
        );
}
```

oandregal replied:
See thread above https://github.com/WordPress/gutenberg/pull/38039#discussion_r808702086

oandregal committed:
New settings (works in WordPress 5.9)

Rename wp_get_global_settings to gutenberg_get_global_settings.

In WordPress 5.9 the wp_* function is already defined,
so we can't override them. It's calling the existing
WP_Theme_JSON classes in WordPress core so it won't pick up
the plugin modification.

In the plugin, to make sure these changes work fine in 5.9
as well, we need to use the gutenberg_* function instead.

oandregal committed:
New styles (works in WordPress 5.8)

oandregal committed:
New styles (works in WordPress 5.9)

andrewserong committed:
Try moving position support to layout

andrewserong committed:
Try: Add sticky position support for the Group block

oandregal committed:
New settings (works in WordPress 5.8)

oandregal committed:
Fix for WordPress 5.9: gutenberg_get_global_stylesheet should call the Gutenberg classes

oandregal committed:
Fix lint issues

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -30,6 +71,98 @@ class WP_Theme_JSON_Gutenberg extends WP_Theme_JSON_5_9 {
+		'custom'          => null,
+		'layout'          => array(
+			'contentSize' => null,
+			'position'    => null,
```
Does this need to be added to `schemas/json/theme.json` as well?

Asking for me :)

E.g.,

```json
		"settingsPropertiesLayout": {
			"properties": {
				"layout": {
					"description": "Settings related to layout.",
					"type": "object",
					"properties": {
						"position": {
							"description": "Determines the sticky position of a block.",
							"type": "string"
						},
						"contentSize": {
							"description": "Sets the max-width of the content.",
							"type": "string"
						},
						"wideSize": {
							"description": "Sets the max-width of wide (`.alignwide`) content.",
							"type": "string"
						}
					},
					"additionalProperties": false
				}
			}
		},
```

andrewserong replied:
Yes, I'm pretty sure it will be! There's still a few other things I wanted to explore in this PR before adding that (like where to put `top`, `right`, `bottom`, `left` in the object, and whether we also need to store a `zIndex` value so that the sticky position will work — even if we keep that setting hidden from the user in the UI)

andrewserong commented:
Update after a very long time away from this PR — the original implementation here was imagining storing the `position` value in the inline style for the block. Now that the layout block support and style engine have progressed considerably, I think there's a good opportunity to rethink how to develop this feature, particularly as this PR uncovered that position requires additional associated styles (such as `top`) and related rules to account for the logged in admin bar.

I'd love to have another go at this from the perspective of the current state of layout. I think some good things to try would be:

* Make `position` be an attribute of layout, and use `layout.php` to output the position based styles.
* `position` would not feature as part of theme.json styling, as a particular type of block should not receive positioning globally — it should be something that is set at the individual block level for a _particular_ group block, but not for _all_ group blocks.
* Therefore, output of `position` rules likely makes sense to be included in either `layout.php` or as part of base layout styles for a particular layout type. Given how infrequently `position` will be used in terms of number of blocks on a site, I'm leaning toward thinking it should be handled in `layout.php` at render time (or attached via a semantic classname).
* Handling the rules at render time would allow us to inject related rules, such as whichever `top`, `bottom`, or `z-index` rules are required for the sticky positioned block to be rendered correctly.

andrewserong commented:
Closing out this experiment now, but let's follow-up separately with new explorations into how to achieve this feature.
